In [1]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset('diamonds')


In [2]:
import pandas as pd
import seaborn as sns
import numpy as np
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

df = sns.load_dataset('diamonds')

df = df[(df[['x', 'y', 'z']] != 0).all(axis=1)]
df_model = pd.get_dummies(df, columns=['cut', 'color', 'clarity'], drop_first=True)
X = df_model.drop('price', axis=1)
y = df_model['price']


In [3]:
from sklearn.feature_selection import SelectKBest, f_regression

selector = SelectKBest(score_func=f_regression, k=10)
X_new = selector.fit_transform(X, y)

mask = selector.get_support()
selected_features = X.columns[mask]
print("Variables seleccionadas:")
print(selected_features)


Variables seleccionadas:
Index(['carat', 'table', 'x', 'y', 'z', 'cut_Premium', 'color_E', 'color_I',
       'clarity_VVS1', 'clarity_SI2'],
      dtype='object')


In [5]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
X_selected = X[selected_features]

X_train, X_test, y_train, y_test = train_test_split(X_selected, y, test_size=0.2, random_state=42)

modelo = LinearRegression()
modelo.fit(X_train, y_train)

y_pred = modelo.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R² Score: {r2:.4f}")


Mean Squared Error (MSE): 2327767.24
R² Score: 0.8548


## Conclusiones

Se entrenó un modelo de regresión lineal para predecir el precio de diamantes utilizando un conjunto reducido de 10 variables seleccionadas mediante el método `SelectKBest`. Entre las variables más influyentes se destacan `carat`, `x`, `y`, `z` y variables categóricas como `clarity_VVS1` y `color_E`.

El modelo alcanzó un **R² de 0.8548**, lo que indica que explica un 85,5% de la variabilidad del precio. Esto sugiere una buena capacidad predictiva dentro del marco de una regresión lineal.

El **MSE fue de 2.327.767**, valor razonable dada la presencia de outliers y la magnitud de los precios.

Como trabajo futuro, podrían explorarse modelos más complejos como `Random Forest` o `XGBoost`, y aplicar técnicas de escalado, normalización y manejo avanzado de outliers para mejorar la performance.
